### PR TITLE
RIOT: auto execute LFC on build

### PR DESCRIPTION
**Pull request by LasseRosenow**
This adds the lfc command to `riot.mk` and also moves more configuration over from the lingo template.

For more info see: https://github.com/lf-lang/lf-riot-template/pull/1

### CODE REVIEW

File: make/riot/riot.mk [hunk 1]
```diff
@@ -1,12 +1,31 @@
...
+# Check if REACTOR_UC_PATH variable exist
+ifndef REACTOR_UC_PATH
+$(error REACTOR_UC_PATH is not defined. Please define it!)
+endif
 
 # Name of your RIOT application
 APPLICATION ?= $(LF_MAIN)
 
+# This has to be the absolute path to the RIOT base directory:
+RIOTBASE ?= $(CURDIR)/RIOT
```
Comment (erlingrj) [inline make/riot/riot.mk:21]: This assumes that you are building your application from our lf-riot-template and have RIOT as a submodule. Ideally we would keep such assumptions in the makefile within `lf-riot-template`. IIRC we discussed moving some of this back to the makefile in `lf-riot-template`, did you request my review before or after the meeting :stuck_out_tongue:

Comment (LasseRosenow) [reply make/riot/riot.mk:21]: Yeah that was before the meeting haha :)

Comment (LasseRosenow) [reply make/riot/riot.mk]: Done

**Comment by LasseRosenow** [comment]
@erlingrj Ready for review again :)

File: examples/riot/buildAll.sh [hunk 2]
```diff
@@ -21,8 +21,3 @@ for board in "${BOARDS[@]}"; do
 			popd
 	done
 done
-
-# Build lf example
-pushd hello_lf
-run/build.sh
-popd
```
Comment (LasseRosenow) [inline examples/riot/buildAll.sh:28]: Integrating the LFC into the make file is really nice! Now we don't need to handle _LF examples different from c-code based examples anymore!

File: examples/riot/hello_lf/Makefile [hunk 3]
```diff
@@ -1,14 +1,20 @@
...
+
+# Enable reactor-uc features
+# CFLAGS += -DNETWORK_CHANNEL_TCP_POSIX
+# CFLAGS += -DNETWORK_CHANNEL_COAP_RIOT
 
-# Path of generated lf c-code
-LF_SRC_GEN_PATH = $(CURDIR)/src-gen/$(LF_MAIN)
+# Execute the LF compiler if build target is "all"
+ifeq ($(MAKECMDGOALS),all)
+  _ :=  $(shell $(REACTOR_UC_PATH)/lfc/bin/lfc-dev src/$(LF_MAIN).lf)
```
Comment (erlingrj) [inline examples/riot/hello_lf/Makefile:12]: Just for my own understanding, what does this syntax here achieve? Will the command be invoked before any target is run?

File: make/riot/riot-lfc.mk [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# Check if required environment variables exist
+ifndef LF_MAIN
+  $(error LF_MAIN is not defined. Please define it!)
+endif
+
+# Name of your RIOT application
+APPLICATION ?= $(LF_MAIN)
+
+# Path of generated lf c-code
+LF_SRC_GEN_PATH ?= $(CURDIR)/src-gen/$(LF_MAIN)
```
Comment (erlingrj) [inline make/riot/riot-lfc.mk:12]: An unfortunate detail about lfc is that where the generated-code is put is a little complicated:
1. `~/src/Hello.lf` -> `~/src-gen/Hello`
2. `~/Hello.lf` -> `~/src-gen/Hello`
3. `~/src/lib/Hello.lf` -> `~/src-gen/lib/Hello

I think we should ignore (2) where it is not in a `src` directory, for our template we say that the LF files must reside under a `src` directory. But for (3) to work, LF_MAIN must be set to `lib/Hello`, which might actually be reasonable

File: make/riot/riot-lfc.mk [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+# Include generated c files
+SRC += $(patsubst %, $(LF_SRC_GEN_PATH)/%, $(LFC_GEN_SOURCES))
+
+# Include generated main file
+SRC += $(LF_SRC_GEN_PATH)/${LFC_GEN_MAIN}
+
+# Include generated h files
+CFLAGS += -I$(LF_SRC_GEN_PATH)
+
+include $(RIOT_MK_DIR)/riot.mk
```
Comment (erlingrj) [inline make/riot/riot-lfc.mk:26]:
```suggestion
include $(RIOT_MK_DIR)/riot.mk

```

**Review by erlingrj — APPROVED** [review]
This looks good. Just a few minor comments. I will let you merge it when you've seen them

File: examples/riot/hello_lf/Makefile (hunk 3 above)
Comment (LasseRosenow) [reply examples/riot/hello_lf/Makefile:12]: This is a small hack, it copies the output of the shell script over to the variable in this case underscore. This way the evaluation of the makefile is blocked until the script is finished ;)

File: make/riot/riot-lfc.mk (hunk 4 above)
Comment (LasseRosenow) [reply make/riot/riot-lfc.mk:12]: Aaahh that is good to know, I will think about a good solution

File: examples/riot/hello_lf/Makefile (hunk 3 above)
Comment (erlingrj) [reply examples/riot/hello_lf/Makefile:12]: That's very clever!

File: make/riot/riot-lfc.mk (hunk 5 above)
Comment (LasseRosenow) [reply make/riot/riot-lfc.mk:26]: GitHub mobile doesn't let me approve this there is a strange error message. Feel free to just commit it :)

**Pull request merged**